test/self_protocore/upsert_validator.js

```diff
@@ -0,0 +1,151 @@
+  const accountProvider = new AccountProvider(accounts);
+
+  beforeEach(async () => {
+    config.coconsensusAddress = accountProvider.get();
+
+    config.epochLength = new BN(100);
+
+    config.genesisKernelHeight = new BN(1);
+    config.genesisKernelHash = Utils.getRandomHash();
+
+    config.core = accountProvider.get();
+    config.genesis = {};
+    config.genesis.auxiliaryParentVoteMessageHash = Utils.getRandomHash();
+    config.genesis.auxiliarySourceTransitionHash = Utils.getRandomHash();
+    config.genesis.auxiliarySourceBlockHash = Utils.getRandomHash();
+    config.genesis.auxiliaryTargetBlockHash = Utils.getRandomHash();
+    config.genesis.auxiliaryAccumulatedGas = new BN(1000000);
+
+    config.setupParams = {};
+    config.setupParams.metachainId = Utils.getRandomHash();
+    config.setupParams.domainSeparator = Utils.getRandomHash();
+    config.setupParams.epochLength = new BN(100);
+    config.setupParams.metablockHeight = new BN(Utils.getRandomNumber(1000));
+    config.setupParams.selfProtocore = accountProvider.get();
+    config.setupParams.coconsensus = accountProvider.get();
+
+    config.genesis.auxiliarySourceBlockNumber = new BN(
+      Utils.getRandomNumber(10000) * config.setupParams.epochLength,
+    );
+    config.genesis.auxiliaryTargetBlockNumber = config.genesis.auxiliarySourceBlockNumber.add(
+      config.setupParams.epochLength,
+    );
+
+    // Deploy the self protocore contract.
+    config.selfProtocore = await TestSelfProtocore.new();
+
+    await config.selfProtocore.setGenesisStorage(
+      config.genesis.auxiliaryParentVoteMessageHash,
+      config.genesis.auxiliarySourceTransitionHash,
+      config.genesis.auxiliarySourceBlockHash,
+      config.genesis.auxiliarySourceBlockNumber,
+      config.genesis.auxiliaryTargetBlockHash,
+      config.genesis.auxiliaryTargetBlockNumber,
+      config.genesis.auxiliaryAccumulatedGas,
+    );
+
+    validator.address = accountProvider.get();
+    validator.beginHeight = new BN(20);
+    validator.endHeight = new BN(100);
+    validator.reputation = new BN(500);
+
+    await config.selfProtocore.setCoconsensus(config.coconsensusAddress);
+    await config.selfProtocore.setOpenKernelHeight(validator.beginHeight);
+
+    await config.selfProtocore.setup(
+      config.setupParams.metachainId,
+      Utils.getRandomHash(),
+      new BN(100),
+      new BN(20),
+      {
+        from: config.coconsensusAddress,
+      },
+    );
+  });
+
+  contract('Positive Tests', async () => {
+    it('should insert new validator if already not present', async () => {
+      await config.selfProtocore.upsertValidator(
+        validator.address,
+        validator.beginHeight,
+        validator.reputation,
+        {
+          from: config.coconsensusAddress,
+        },
+      );
+
+      assert.notStrictEqual(
+        await config.selfProtocore.validators.call(validator.address),
+        Utils.NULL_ADDRESS,
+        'Expected address is SENTINEL_ADDRESS(0x1)',
```

Comment:
When we know what should be the expected result, then why don't we do a strict check?
Either change it to `strictEqual` or change the reason that it should not be a null address.

Reply:
Correct. We know the expected result.
But it is done as per the discussion at https://github.com/mosaicdao/mosaic-1/pull/212#discussion_r370044049